### PR TITLE
feat: broker can encode btc smart contract call

### PR DESCRIPTION
**Pull request by j4m1ef0rd**
# Pull Request

Closes: PRO-1648

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [ ] I have written sufficient tests.
- [ ] I have written and tested required migrations.
- [ ] I have updated documentation where appropriate.

## Summary

- new `broker_encodeBtcSmartContractCall` rpc that will create an encoded btc start contract call with the given params.
- Extended the `SwapLimitsProvider` trait to validate the dca and refund params.
	- Added an RPC to use this param validation so that the broker api can use it. 

### TODO
- Documentation

### CODE REVIEW

File: api/lib/src/lib.rs [hunk 1]
```diff
@@ -515,6 +523,72 @@ pub trait BrokerApi: SignedExtrinsicApi + StorageApi + Sized + Send + Sync + 'st
 		self.simple_submission_with_dry_run(pallet_cf_swapping::Call::deregister_as_broker {})
 			.await
 	}
+
+	async fn encode_btc_smart_contract_call(
```
Comment (CumpsD) [inline api/lib/src/lib.rs:527]: Does this take brokers and affiliate brokers into account too?

Comment (dandanlen) [reply api/lib/src/lib.rs:527]: Not yet. For btc it's not trivial, will be added separately.

File: api/bin/chainflip-broker-api/src/main.rs [hunk 2]
```diff
@@ -48,6 +48,17 @@ pub trait Rpc {
...
+	#[method(name = "encode_btc_smart_contract_call", aliases = ["broker_encodeBtcSmartContractCall"])]
+	async fn encode_btc_smart_contract_call(
+		&self,
+		output_asset: Asset,
+		output_address: AddressString,
+		retry_duration: u32,
+		min_output_amount: u128,
+		boost_fee: Option<BasisPoints>,
+		dca_parameters: Option<DcaParameters>,
+	) -> RpcResult<Vec<u8>>;
```
Comment (dandanlen) [inline api/bin/chainflip-broker-api/src/main.rs:61]: Vec<u8> is not a very json-friendly return value. At the very least it should be hex-encoded, but I think we can also do better to communicate *what* it is. For example, we could make it an object with a named field. Combined with the above suggestion, a suitable output type might be:

```
pub enum SwapPayload {
	Bitcoin {
		nulldata_utxo: Bytes, // sp_core::Bytes serializes as hex
	},
	// Other variants to be added later: Eth, Sol
}
```

File: api/lib/src/lib.rs [hunk 3]
```diff
@@ -321,8 +329,8 @@ impl AddressString {
 			.map_err(|_| anyhow!("Failed to parse address"))
 	}
 
-	pub fn from_encoded_address(address: &EncodedAddress) -> Self {
-		Self(address.to_string())
+	pub fn from_encoded_address<T: std::borrow::Borrow<EncodedAddress>>(address: T) -> Self {
+		Self(address.borrow().to_string())
```
Comment (dandanlen) [inline api/lib/src/lib.rs:333]: What's the reason for this change?

File: api/lib/src/lib.rs (hunk 1 above)
Comment (CumpsD) [reply api/lib/src/lib.rs:527]: > Not yet. For btc it's not trivial, will be added separately.

Ok, so to start I will just provide this functionality via BaaS, and if I am correct, BTC -> x swaps initiated this way will start out with 0% commission to the affiliate (and main broker), right?

File: api/lib/src/lib.rs [hunk 4]
```diff
@@ -515,6 +523,72 @@ pub trait BrokerApi: SignedExtrinsicApi + StorageApi + Sized + Send + Sync + 'st
...
+				}
+				let total_swap_request_duration = params
+					.number_of_chunks
+					.saturating_sub(1)
+					.checked_mul(params.chunk_interval)
+					.ok_or(anyhow!("Invalid DCA parameters"))?;
+
+				if total_swap_request_duration > max_swap_request_duration_blocks {
+					bail!("Invalid DCA parameters: Total swap request duration must be <= {max_swap_request_duration_blocks} blocks");
+				}
```
Comment (dandanlen) [inline api/lib/src/lib.rs:567]: This is a duplication of logic form the ingress pallet - would be good to put this in shared function, otherwise it's easy to introduce inconsistencies.

File: api/lib/src/lib.rs [hunk 5]
```diff
@@ -515,6 +523,72 @@ pub trait BrokerApi: SignedExtrinsicApi + StorageApi + Sized + Send + Sync + 'st
...
+					.try_into()?,
+				chunk_interval: dca_parameters
+					.as_ref()
+					.map(|params| params.chunk_interval)
+					.unwrap_or(2)
+					.try_into()?,
+				boost_fee: boost_fee.unwrap_or_default().try_into()?,
+			},
+		};
+		Ok(encode_swap_params_in_nulldata_utxo(params))
```
Comment (dandanlen) [inline api/lib/src/lib.rs:590]: Might be worth clarifying if we want to return the encoded utxo, or just the data (ie. whether we include the OP_RETURN or not).

File: api/lib/src/lib.rs (hunk 1 above)
Comment (dandanlen) [reply api/lib/src/lib.rs:527]: Adding broker fees is a separate issue from what is being added here. By the time release this, we may also have support for broker/affiliate fees.

Having said that, there is no real reason to prefer this method over the other unless for whatever reason you don't want to wait 6secs to open a deposit channel.

File: api/bin/chainflip-broker-api/src/main.rs [hunk 6]
```diff
@@ -48,6 +48,17 @@ pub trait Rpc {
 		asset: Asset,
 		destination_address: AddressString,
 	) -> RpcResult<WithdrawFeesDetail>;
+
+	#[method(name = "encode_btc_smart_contract_call", aliases = ["broker_encodeBtcSmartContractCall"])]
+	async fn encode_btc_smart_contract_call(
```
[the remark this answers is not in the document]
Comment (CumpsD) [reply api/bin/chainflip-broker-api/src/main.rs:53]: `sign_swap_payload`?

[the remark this answers is not in the document]
Comment (msgmaxim) [reply api/bin/chainflip-broker-api/src/main.rs:53]: How about `request_swap_parameter_encoding`?

File: api/lib/src/lib.rs (hunk 5 above)
Comment (msgmaxim) [reply api/lib/src/lib.rs:590]: IMO returning just the data makes more sense. Integratorss might feel more in control if they are the ones who create the actual UTXO (e.g. ensuring that it is non-spendable etc). Also, this is going to be more consistent with other chains when we add them.

File: api/lib/src/lib.rs (hunk 3 above)
Comment (j4m1ef0rd) [reply api/lib/src/lib.rs:333]: Just to fix rust analyser complaining about it. If i just removed the reference then clippy would complain about it, so this was the best solution that made both happy.

File: api/bin/chainflip-broker-api/src/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (j4m1ef0rd) [reply api/bin/chainflip-broker-api/src/main.rs:53]: > it should be chain-agnostic

How does the user specify which encoding they want? should I add a parameter for the encodeing type/chain?
or just return all encoding types for the swap?

File: api/lib/src/lib.rs (hunk 1 above)
Comment (kylezs) [inline api/lib/src/lib.rs:527]: Thinking, could we add a safe mode param for this? So if safe mode is enabled, this query doesn't work. It's not bullet proof since people could construct it on their own without using this - but more than likely people will be using this, and we can prevent the creation of these memo calls

File: api/bin/chainflip-broker-api/src/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dandanlen) [reply api/bin/chainflip-broker-api/src/main.rs:53]: The input asset determines the chain. The output can be an enum with one variant per chain.

**Review by msgmaxim — COMMENTED** [review]
Looks good (apart from things that Dan already commented on)

File: api/lib/src/lib.rs [hunk 7]
```diff
@@ -515,6 +528,75 @@ pub trait BrokerApi: SignedExtrinsicApi + StorageApi + Sized + Send + Sync + 'st
...
+			bail!("Safe mode is active. Swaps are disabled.");
+		}
+
+		// Validate params
+		let params = serde_json::to_string(&RefundValidationParams { retry_duration })?;
+		let raw_value = serde_json::value::RawValue::from_string(params)?;
+		self.base_rpc_api()
+			.request_raw("cf_validate_refund_params", Some(raw_value))
+			.await
+			.map_err(anyhow::Error::msg)?;
```
[the remark this answers is not in the document]
Comment (j4m1ef0rd) [reply api/lib/src/lib.rs:559]: I couldnt find a way to use them directly from the `BrokerApi` so i added them to the `BaseRpcApi`.

[the remark this answers is not in the document]
Comment (dandanlen) [reply api/lib/src/lib.rs]: Cool - would be good to find a nice way to expose the api directly instead of having to add this boilerplate... can wait for another time though.

**Comment by dandanlen** [comment]
Adding the broker fees is a nice idea but we should not just silently ignore them - better to return an error message for now.

**Comment by dandanlen** [comment]
> Adding the broker fees is a nice idea but we should not just silently ignore them - better to return an error message for now.

☝️ Will add this and then merge, thanks @j4m1ef0rd

**Comment by dandanlen** [comment]
Actually, I'll also rename the params so that they match `request_swap_deposit_address` as closely as possible - will make documentation much easier.

**Review by dandanlen — APPROVED** [review]

**Comment by CumpsD** [comment]
Any reason you didnt keep this (https://github.com/chainflip-io/chainflip-backend/blob/main/api/bin/chainflip-broker-api/src/main.rs#L61-L65) consistent with the order of https://github.com/chainflip-io/chainflip-backend/blob/main/api/bin/chainflip-broker-api/src/main.rs#L41-L44 ?

**Comment by CumpsD** [comment]
Also, my Rust is rusty, but here you got `affiliate_fees.unwrap_or_default(),` https://github.com/chainflip-io/chainflip-backend/blob/main/api/bin/chainflip-broker-api/src/main.rs#L119C5-L119C40

While here you dont: https://github.com/chainflip-io/chainflip-backend/blob/main/api/bin/chainflip-broker-api/src/main.rs#L164

**Comment by CumpsD** [comment]
Out of curiosity, how are refunds handled in this way? Since there is no refund address, where do you get it from?

**Comment by j4m1ef0rd** [comment]
> Out of curiosity, how are refunds handled in this way? Since there is no refund address, where do you get it from?

The refund address is stored in the 3rd output of the bitcoin transaction. Will be explained in the docs.

> but here you got affiliate_fees.unwrap_or_default()

We would just unwrao_or_default in the next function. Does not really matter at what point it becomes non optional. You're right, we should be consistent.

> Any reason you didnt keep this consistent with the order of

At first it was because I wanted to put the optional params at the end also some of the params where different, so it could not be exactly the same, but Dan changed it. So now there is no reason they are not consistent 😅.

**Comment by dandanlen** [comment]
Re. consistency:

I tried to keep it as consistent as possible. 

There are some necessary differences between requesting a deposit channel vs. requesting a transaction payload. 

The specific reason for the change in order in this case is that retry_duration and min_output_amount are non-optional and replace `Option<RefundParameters>` from the first method. As @j4m1ef0rd mentioned, we keep optional parameters to the end because when passing the parameters in an array, having optional params in the middle of the array creates some weird ambiguities (should it be `null`? should it be left out? etc.). 

Unifying this would have required changing the original api so I opted not to (for now). My main concern was for naming. 

In general, however, I would highly recommend using on object rather than an array for passing the parameters, then the order is a non-issue.

**Comment by CumpsD** [comment]
Object it is! thanks
